firefox-ios/Client/Frontend/Browser/DownloadQueue.swift

```diff
@@ -147,6 +159,30 @@ class HTTPDownload: Download {
             self.task?.resume()
         }
     }
+
+    // Makes a HEAD request to check the unencoded content size of the file, which reflects the NSURLSession delivery bytes
+    // size given via the `URLSessionDownloadDelegate` methods (because decoding is automatic).
+    // Note: `NSURLSession` automatically inserts the following header into requests, which can result in encoded
+    // (compressed) data from a server: `Accept-Encoding: gzip;q=1.0, compress;q=0.5`
+    func getUncompressedSize() async -> Int64? {
+        if let url = self.request.url {
+            var headRequest = URLRequest(url: url)
+            headRequest.httpMethod = "HEAD"
+            // Override URLSession's default Accept-Encoding headers (`gzip;q=1.0, compress;q=0.5`)
+            // to instead accept no transformation of the file at all.
+            headRequest.setValue("identity", forHTTPHeaderField: "Accept-Encoding")
+
```

Comment:
This seems Ok to me, though it also seems surprising that we'd have to do this (I am not an expert in how the encoding/gzip is handled for HTTP requests, so I don't know offhand if there's a better solution).

Do you recall if this issue is only happening on iOS? (IIRC it doesn't replicate on Firefox Android / Mac etc?) If that's the case it might not hurt to reach out to those teams to see what approach they're using for their download progress bars for gzipped payloads. 🤔

Reply:
I hear you! I was really hoping to do better than simply provide a workaround. 🫤  But from what I could tell, it appears that iOS simply does not offer the option to _not_ automatically decode compressed bytes, which basically means our hands are tied. (I actually would love another pair of eyes on this just to see if someone else comes to the same conclusion or finds something I missed).

Here's an [open radar ticket from 2016](http://openradar.appspot.com/radar?id=4931101243998208) requesting a way to disable automatic HTTP content decoding.

Here someone apparently leverages a [private API](https://stackoverflow.com/a/16705063) to disable the automatic behaviour...? This could have promise, but I wasn't sure there was a feasible way to test this with our setup (it's getting a bit outside my wheelhouse).

---

Good idea to test what the other Firefox products are doing!

I just tested the android app and there is no UI for downloads. The download goes into the system notification area.

Firefox for desktop (from the Downloads Library) shows the compressed bytes, and then when the download finishes, you see the uncompressed bytes.
<img width="800" alt="Screenshot 2024-07-29 at 1 37 48 PM" src="https://github.com/user-attachments/assets/554c7eb5-f62f-4e74-a05d-c958b05e9da3">
<img width="800" alt="Screenshot 2024-07-29 at 1 38 50 PM" src="https://github.com/user-attachments/assets/fd44632a-4fcb-4e5a-b5c5-a6d05f53bcb5">

For another iOS comparison, Safari shows very little feedback. Just that the file is downloading, and then fully downloaded. No progress tracking. Watch the little bubble that appears beside the address bar in the video below:

https://github.com/user-attachments/assets/5809c94a-8709-4946-a045-3280e9d69fdd

Maybe we should go the route of Safari!

Reply:
>  Safari shows very little feedback. Just that the file is downloading, and then fully downloaded. No progress tracking. 

I think that might only be on iPhone devices, because at least on iPad Safari appears to have a progress tracker:

<img width="291" alt="Screenshot 2024-07-29 at 1 15 33 PM" src="https://github.com/user-attachments/assets/fc54b3c8-900b-4573-9c02-d1b20a7a0a8a">

Reply:
> It sees that it's gzip I guess and just says "?"...

I wonder if we should consider doing something similar? I think that would be a product decision ultimately so we'd probably want to check with Andy & UX folks. Thanks again for all of the digging you did into this issue @ih-codes

Reply:
I think that would be a good compromise. I'll reach out to Andy and see where we should go from here. Thanks for reviewing and pointing me toward the right places @mattreaganmozilla! 👍